examples/fabric-admin/device_manager/DeviceManager.cpp

```diff
@@ -0,0 +1,226 @@
+            return const_cast<Device *>(&device);
+        }
+    }
+    return nullptr;
+}
+
+void DeviceManager::RemoveSyncedDevice(NodeId nodeId)
+{
+    Device * device = FindDeviceByNode(nodeId);
+    if (device != nullptr)
+    {
+        mSyncedDevices.erase(*device);
+        ChipLogProgress(NotSpecified, "Removed synced device: NodeId:" ChipLogFormatX64 ", EndpointId %u",
+                        ChipLogValueX64(device->GetNodeId()), device->GetEndpointId());
+    }
+    else
+    {
+        ChipLogProgress(NotSpecified, "No device found with NodeId:" ChipLogFormatX64, ChipLogValueX64(nodeId));
+    }
+}
+
+void DeviceManager::HanldeAttributeChange(const app::ConcreteDataAttributePath & path, TLV::TLVReader * data)
+{
+    if (path.mClusterId != Descriptor::Id || path.mAttributeId != Descriptor::Attributes::PartsList::Id)
+    {
+        return;
+    }
+
+    ChipLogProgress(NotSpecified, "Attribute change detected:");
+    ChipLogProgress(
+        NotSpecified, "Endpoint: %u, Cluster: " ChipLogFormatMEI ", Attribute: " ChipLogFormatMEI ", DataVersion: %" PRIu32,
+        path.mEndpointId, ChipLogValueMEI(path.mClusterId), ChipLogValueMEI(path.mAttributeId), path.mDataVersion.ValueOr(0));
+
+    app::DataModel::DecodableList<EndpointId> value;
+    CHIP_ERROR error = app::DataModel::Decode(*data, value);
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to decode attribute value. Error: %" CHIP_ERROR_FORMAT, error.Format());
+        return;
+    }
+
+    std::set<EndpointId> newEndpoints;
+
+    // Populate the newEndpoints set from the decoded value using an iterator
+    auto iter = value.begin();
+    while (iter.Next())
+    {
+        newEndpoints.insert(iter.GetValue());
+    }
+
+    if (iter.GetStatus() != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to iterate over decoded attribute value.");
+        return;
+    }
+
+    // Compare newEndpoints with mSyncedEndpoints to determine added and removed endpoints
+    std::vector<EndpointId> addedEndpoints;
+    std::vector<EndpointId> removedEndpoints;
+
+    // Find added endpoints
+    for (const auto & endpoint : newEndpoints)
+    {
+        if (FindDeviceByEndpoint(endpoint) == nullptr)
+        {
+            addedEndpoints.push_back(endpoint);
+        }
+    }
+
+    // Find removed endpoints
+    for (auto & device : mSyncedDevices)
+    {
+        EndpointId endpointId = device.GetEndpointId();
+        if (newEndpoints.find(endpointId) == newEndpoints.end())
+        {
+            removedEndpoints.push_back(endpointId);
+        }
+    }
+
+    // Process added endpoints
+    for (const auto & endpoint : addedEndpoints)
+    {
+        ChipLogProgress(NotSpecified, "Endpoint added: %u", endpoint);
+
+        if (mAutoSyncEnabled)
+        {
+            char command[64];
+            snprintf(command, sizeof(command), "fabricsync sync-device %d", endpoint);
+
+            usleep(kSyncPrepareTimeMs * 1000);
```

Comment:
What thread is this sleeping on? If matter thread can we not schedule this?

Reply:
It is on matter thread, maybe you are right, we don't need delay here since it is in sequence.